### PR TITLE
Replace deprecated Controller by AbstractController

### PR DESCRIPTION
**Pull request by chindit**
## Replace deprecated Controller by AbstractController

In Symfony 4.2, `Controller` class is deprecated in favor of `AbstractController`.
Changes are really small as the methods are almost identical.
Symfony changelog is available here: https://symfony.com/blog/new-in-symfony-4-2-important-deprecations

I am targeting this branch, because it does imply BC-break

## Changelog

```markdown

### Changed
- Usages of `Symfony\Bundle\FrameworkBundle\Controller\Controller` removed in favor of Dependency Injection
```

### CODE REVIEW

File: src/Action/DashboardAction.php [hunk 1]
```diff
@@ -16,11 +16,11 @@
...
 use Sonata\AdminBundle\Admin\Pool;
 use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
-use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 
-final class DashboardAction extends Controller
+final class DashboardAction extends AbstractController
 {
```
Comment (core23) [inline src/Action/DashboardAction.php]: IMHO we should not refer to any controller, we should register controller as services

**Comment by OskarStark** [comment]
As we support Symfony 3, this change is not possible in `3.x` branch IMO

**Comment by chindit** [comment]
> As we support Symfony 3, this change is not possible in `3.x` branch IMO

I wasn't sure on which branch make this PR.  I will change to `master` in this case.

**Comment by OskarStark** [comment]
Please check the travis build @chindit

File: src/Controller/CRUDController.php [hunk 2]
```diff
@@ -1525,6 +1525,11 @@ protected function preList(Request $request)
         return $this->get('translator')->trans($id, $parameters, $domain, $locale);
     }
 
+    private function getSelectedTab(Request $request)
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]:
```suggestion
    private function getSelectedTab(Request $request): array
```

File: src/Action/DashboardAction.php [hunk 3]
```diff
@@ -16,11 +16,11 @@
 use Sonata\AdminBundle\Admin\BreadcrumbsBuilderInterface;
 use Sonata\AdminBundle\Admin\Pool;
 use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
-use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
 
-final class DashboardAction extends Controller
+final class DashboardAction extends AbstractController
```
Comment (greg0ire) [inline src/Action/DashboardAction.php]: A better goal IMO would be to completely remove the `extends AbstractController` part, and use dependency injection.

Comment (OskarStark) [reply src/Action/DashboardAction.php]: See https://symfony.com/doc/current/controller/service.html for more information.

File: src/Action/DashboardAction.php [hunk 4]
```diff
@@ -41,17 +42,23 @@ final class DashboardAction extends Controller
...
+     */
+    private $templatingEngine;
 
     public function __construct(
         array $dashboardBlocks,
         BreadcrumbsBuilderInterface $breadcrumbsBuilder,
         TemplateRegistryInterface $templateRegistry,
-        Pool $pool
+        Pool $pool,
+        EngineInterface $templatingEngine
```
Comment (OskarStark) [inline src/Action/DashboardAction.php]: Should be templateEngine

File: src/Action/SearchAction.php [hunk 5]
```diff
@@ -45,39 +45,45 @@ final class SearchAction extends Controller
...
         $this->breadcrumbsBuilder = $breadcrumbsBuilder;
+        $this->templatingEngine = $templatingEngine;
     }
 
     /**
      * The search action first render an empty page, if the query is set, then the template generates
      * some ajax request to retrieve results for each admin. The Ajax query returns a JSON response.
      *
-     * @throws \RuntimeException
+     * @param Request $request
```
Comment (OskarStark) [inline src/Action/SearchAction.php]: Can be removed

File: src/Action/DashboardAction.php (hunk 4 above)
Comment (chindit) [reply src/Action/DashboardAction.php]: Okay, done

File: composer.json [hunk 6]
```diff
@@ -63,6 +63,7 @@
         "jms/di-extra-bundle": "^1.9",
         "jms/translation-bundle": "^1.4",
         "matthiasnoback/symfony-dependency-injection-test": "^3.1",
+        "phpspec/prophecy": "1.8.*",
```
Comment (phansys) [inline composer.json]: This change should be done in a separate PR, and the constraint should trust semver.

Comment (OskarStark) [reply composer.json]: Yes please use ^1.8

**Comment by VincentLanglet** [comment]
> As we support Symfony 3, this change is not possible in `3.x` branch IMO

@OskarStark What's making you say this change is not possible in `3.x` ?
I use `AbstractController` in Symfony 3 so I don't think this PR needs a major release.

**Comment by phansys** [comment]
Class `AbstractController` doesn't extend the deprecated class `Controller`, so it represents a break for our BC promise given the current API.

See https://github.com/symfony/symfony/blob/v4.2.0/UPGRADE-4.2.md#frameworkbundle.

**Comment by VincentLanglet** [comment]
@phansys What about splitting the PR ?
Extending class `AbstractController` only for the final class should not be a break for your BC promise, isn't it ?

**Comment by phansys** [comment]
I think yes, that can be done for final classes since that should not break BC at all :+1:
